File: Array/34_find_first_and_last_position_of_element_in_sorted_array.cpp
class Solution {
private:
    int helper(vector<int>& nums, int target, int search, int left, int right) {
        int idx = -1;

        while(left <= right) {
            int pivot = left + (right - left) / 2;

            if(target < nums[pivot]) right = pivot - 1;
            else if(target > nums[pivot]) left = pivot + 1;
            else {
                idx = pivot;
                if(search == 0) right = pivot - 1;
                else left = pivot + 1;
            }
        }
        
        return idx;
    }

public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int size = nums.size();
        int left = helper(nums, target, 0, 0, size - 1);
        if(left == -1) return {-1, -1};
        int right = helper(nums, target, 1, left, size - 1);
        return {left, right};
    }
};

File: Sorting/34_find_first_and_last_position_of_element_in_sorted_array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def search(left, right, first):
            index = -1
            while left <= right:
                pivot = left + (right - left) // 2
                
                if nums[pivot] == target:
                    index = pivot
                    if first:
                        right = pivot - 1
                    else:
                        left = pivot + 1
                elif nums[pivot] > target:
                    right = pivot - 1
                else:
                    left = pivot + 1

            return index

        n = len(nums)
        return [search(0, n - 1, True), search(0, n - 1, False)]